ICE: Item being checked wasn't a function/closure

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code
code from `./src/test/ui/feature-gates/feature-gate-abi.rs`
```Rust
trait Tr {
    extern "rust-call" fn fun() {}
}

fn main() {}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.50.0-nightly (1f95c91c8 2020-12-03)
binary: rustc
commit-hash: 1f95c91c887acb2d0c49f549a07025178b818d87
commit-date: 2020-12-03
host: x86_64-unknown-linux-gnu
release: 1.50.0-nightly
```

### Error output

```
error[E0658]: rust-call ABI is subject to change
 --> ./src/test/ui/feature-gates/feature-gate-abi.rs:2:12
  |
2 |     extern "rust-call" fn fun() {}
  |            ^^^^^^^^^^^
  |
  = note: see issue #29625 <https://github.com/rust-lang/rust/issues/29625> for more information
  = help: add `#![feature(unboxed_closures)]` to the crate attributes to enable

error: internal compiler error: compiler/rustc_typeck/src/check/check.rs:108:25: Item being checked wasn't a function/closure: TraitItem(TraitItem { ident: fun#0, hir_id: HirId { owner: DefId(0:4 ~ feature_gate_abi[317d]::Tr::fun), local_id: 0 }, attrs: [], generics: Generics { params: [], where_clause: WhereClause { predicates: [], span: ./src/test/ui/feature-gates/feature-gate-abi.rs:2:32: 2:32 (#0) }, span: ./src/test/ui/feature-gates/feature-gate-abi.rs:2:30: 2:30 (#0) }, kind: Fn(FnSig { header: FnHeader { unsafety: Normal, constness: NotConst, asyncness: NotAsync, abi: RustCall }, decl: FnDecl { inputs: [], output: DefaultReturn(./src/test/ui/feature-gates/feature-gate-abi.rs:2:33: 2:33 (#0)), c_variadic: false, implicit_self: None }, span: ./src/test/ui/feature-gates/feature-gate-abi.rs:2:5: 2:32 (#0) }, Provided(BodyId { hir_id: HirId { owner: DefId(0:4 ~ feature_gate_abi[317d]::Tr::fun), local_id: 2 } })), span: ./src/test/ui/feature-gates/feature-gate-abi.rs:2:5: 2:35 (#0) })

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:958:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.50.0-nightly (1f95c91c8 2020-12-03) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `Tr::fun`
#1 [typeck_item_bodies] type-checking all item bodies
end of query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error[E0658]: rust-call ABI is subject to change
 --> ./src/test/ui/feature-gates/feature-gate-abi.rs:2:12
  |
2 |     extern "rust-call" fn fun() {}
  |            ^^^^^^^^^^^
  |
  = note: see issue #29625 <https://github.com/rust-lang/rust/issues/29625> for more information
  = help: add `#![feature(unboxed_closures)]` to the crate attributes to enable

error: internal compiler error: compiler/rustc_typeck/src/check/check.rs:108:25: Item being checked wasn't a function/closure: TraitItem(TraitItem { ident: fun#0, hir_id: HirId { owner: DefId(0:4 ~ feature_gate_abi[317d]::Tr::fun), local_id: 0 }, attrs: [], generics: Generics { params: [], where_clause: WhereClause { predicates: [], span: ./src/test/ui/feature-gates/feature-gate-abi.rs:2:32: 2:32 (#0) }, span: ./src/test/ui/feature-gates/feature-gate-abi.rs:2:30: 2:30 (#0) }, kind: Fn(FnSig { header: FnHeader { unsafety: Normal, constness: NotConst, asyncness: NotAsync, abi: RustCall }, decl: FnDecl { inputs: [], output: DefaultReturn(./src/test/ui/feature-gates/feature-gate-abi.rs:2:33: 2:33 (#0)), c_variadic: false, implicit_self: None }, span: ./src/test/ui/feature-gates/feature-gate-abi.rs:2:5: 2:32 (#0) }, Provided(BodyId { hir_id: HirId { owner: DefId(0:4 ~ feature_gate_abi[317d]::Tr::fun), local_id: 2 } })), span: ./src/test/ui/feature-gates/feature-gate-abi.rs:2:5: 2:35 (#0) })

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:958:9
stack backtrace:
   0:     0x7f67fb298990 - std::backtrace_rs::backtrace::libunwind::trace::h746c3e9529d524bc
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x7f67fb298990 - std::backtrace_rs::backtrace::trace_unsynchronized::h86340908ff889faa
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f67fb298990 - std::sys_common::backtrace::_print_fmt::h43f85f9b18230404
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7f67fb298990 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc132ae1a5b5aa7cd
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x7f67fb30c58c - core::fmt::write::hdf023a0036d2a25f
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/core/src/fmt/mod.rs:1078:17
   5:     0x7f67fb28a462 - std::io::Write::write_fmt::h8580846154bcb66a
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/io/mod.rs:1519:15
   6:     0x7f67fb29c615 - std::sys_common::backtrace::_print::h7ee55fed88d107a3
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x7f67fb29c615 - std::sys_common::backtrace::print::h54a7d3e52a524177
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x7f67fb29c615 - std::panicking::default_hook::{{closure}}::h60921e857bf55a40
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/panicking.rs:208:50
   9:     0x7f67fb29c16a - std::panicking::default_hook::hf0f9afb1017317fc
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/panicking.rs:225:9
  10:     0x7f67fbb30148 - rustc_driver::report_ice::hff78d76a39ffbb86
  11:     0x7f67fb29cf16 - std::panicking::rust_panic_with_hook::h8d66bf42b407aaea
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/panicking.rs:595:17
  12:     0x7f67fec8d58d - std::panicking::begin_panic::{{closure}}::he3ac55d11a883a10
  13:     0x7f67fec8d296 - std::sys_common::backtrace::__rust_end_short_backtrace::h4402bc3ed558879b
  14:     0x7f67fec8d52f - std::panicking::begin_panic::hd2137c659c375844
  15:     0x7f67fecc79dc - rustc_errors::HandlerInner::bug::hbfb11e3c8ba1475f
  16:     0x7f67fecc60f0 - rustc_errors::Handler::bug::ha00f48e1291906a1
  17:     0x7f67fe563e54 - rustc_middle::util::bug::opt_span_bug_fmt::{{closure}}::h121c3336b55e047b
  18:     0x7f67fe5629db - rustc_middle::ty::context::tls::with_opt::{{closure}}::h61e852c60289ba39
  19:     0x7f67fe562982 - rustc_middle::ty::context::tls::with_opt::h7523943876fa0859
  20:     0x7f67fe563d73 - rustc_middle::util::bug::opt_span_bug_fmt::h131d336df190dd17
  21:     0x7f67fe563ce5 - rustc_middle::util::bug::bug_fmt::hd70ee7a62354ffb4
  22:     0x7f67fc639944 - rustc_typeck::check::check::check_fn::heb12fddb61eb47fe
  23:     0x7f67fc813c86 - rustc_infer::infer::InferCtxtBuilder::enter::h8f9578ee3213fdbe
  24:     0x7f67fc8900b8 - rustc_typeck::check::typeck::h26b1d4f327e46c94
  25:     0x7f67fc7ab9ae - rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::typeck>::compute::h50f9fd8d083760a1
  26:     0x7f67fc76cb3b - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task::h95cf58173480903c
  27:     0x7f67fc77b67b - rustc_data_structures::stack::ensure_sufficient_stack::h0ef6bb96e956214e
  28:     0x7f67fc8ae014 - rustc_query_system::query::plumbing::get_query_impl::h1072b648cee2328c
  29:     0x7f67fc91ccfd - rustc_query_system::query::plumbing::ensure_query_impl::hfe99ea2efbe44936
  30:     0x7f67fc894636 - rustc_typeck::check::typeck_item_bodies::hf28dacb54d1499d4
  31:     0x7f67fc7ac7be - rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::typeck_item_bodies>::compute::h03b6a8389b40a3ef
  32:     0x7f67fc76f024 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_task::haf03bb47557a6c22
  33:     0x7f67fc7859a6 - rustc_data_structures::stack::ensure_sufficient_stack::hca9c5b34c52adf01
  34:     0x7f67fc8bc964 - rustc_query_system::query::plumbing::get_query_impl::h35e55d89fe2c8bdc
  35:     0x7f67fc7d4305 - rustc_typeck::check_crate::hab874d5f99f602e6
  36:     0x7f67fbd8f3f2 - rustc_interface::passes::analysis::h34f9b972d8af6f3d
  37:     0x7f67fbaec79b - rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute::h95ff81a8a73eaff3
  38:     0x7f67fbb7e6e6 - rustc_query_system::dep_graph::graph::DepGraph<K>::with_eval_always_task::h3ea55e2ef223bb62
  39:     0x7f67fbb8dcc6 - rustc_data_structures::stack::ensure_sufficient_stack::ha70725b123b07fd9
  40:     0x7f67fbb1a4b4 - rustc_query_system::query::plumbing::get_query_impl::hcc63ca689fa3c69b
  41:     0x7f67fbaeea3b - rustc_interface::passes::QueryContext::enter::h2071e0a8631ee0e4
  42:     0x7f67fbb8b325 - rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter::hd899306a06575d0c
  43:     0x7f67fbb59217 - rustc_span::with_source_map::ha4e07ff263d0dc1d
  44:     0x7f67fbb870cc - scoped_tls::ScopedKey<T>::set::h39c0aa543118d3f3
  45:     0x7f67fbb8e576 - std::sys_common::backtrace::__rust_begin_short_backtrace::h1e5aa72fb9cd6d86
  46:     0x7f67fbaf67ea - core::ops::function::FnOnce::call_once{{vtable.shim}}::hc793837e985b77ce
  47:     0x7f67fb2ac99a - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hea1090dbdcecbf5a
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/alloc/src/boxed.rs:1327:9
  48:     0x7f67fb2ac99a - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h8d5723d3912bd325
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/alloc/src/boxed.rs:1327:9
  49:     0x7f67fb2ac99a - std::sys::unix::thread::Thread::new::thread_start::hc17a425ca2995724
                               at /rustc/1f95c91c887acb2d0c49f549a07025178b818d87/library/std/src/sys/unix/thread.rs:71:17
  50:     0x7f67fb1a23e9 - start_thread
  51:     0x7f67fb0bf293 - __GI___clone
  52:                0x0 - <unknown>

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.50.0-nightly (1f95c91c8 2020-12-03) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `Tr::fun`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.
```

</p>
</details>



<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"CraftSpider"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_typeck/src/check/check.rs
+++ b/compiler/rustc_typeck/src/check/check.rs
@@ -103,6 +103,9 @@ pub(super) fn check_fn<'a, 'tcx>(
                 Node::ImplItem(hir::ImplItem {
                     kind: hir::ImplItemKind::Fn(header, ..), ..
                 }) => Some(header),
+                Node::TraitItem(hir::TraitItem {
+                    kind: hir::TraitItemKind::Fn(header, .. ), ..
+                }) => Some(header),
                 // Closures are RustCall, but they tuple their arguments, so shouldn't be checked
                 Node::Expr(hir::Expr { kind: hir::ExprKind::Closure(..), .. }) => None,
                 node => bug!("Item being checked wasn't a function/closure: {:?}", node),

--- a/compiler/rustc_typeck/src/check/check.rs
+++ b/compiler/rustc_typeck/src/check/check.rs
@@ -104,7 +104,8 @@ pub(super) fn check_fn<'a, 'tcx>(
                     kind: hir::ImplItemKind::Fn(header, ..), ..
                 }) => Some(header),
                 Node::TraitItem(hir::TraitItem {
-                    kind: hir::TraitItemKind::Fn(header, .. ), ..
+                    kind: hir::TraitItemKind::Fn(header, ..),
+                    ..
                 }) => Some(header),
                 // Closures are RustCall, but they tuple their arguments, so shouldn't be checked
                 Node::Expr(hir::Expr { kind: hir::ExprKind::Closure(..), .. }) => None,

--- a/src/test/ui/abi/issues/issue-22565-rust-call.rs
+++ b/src/test/ui/abi/issues/issue-22565-rust-call.rs
@@ -3,6 +3,31 @@
 extern "rust-call" fn b(_i: i32) {}
 //~^ ERROR A function with the "rust-call" ABI must take a single non-self argument that is a tuple
 
+trait Tr {
+    extern "rust-call" fn a();
+    //~^ ERROR A function with the "rust-call" ABI must take a single non-self argument
+
+    extern "rust-call" fn b() {}
+    //~^ ERROR A function with the "rust-call" ABI must take a single non-self argument
+}
+
+struct Foo;
+
+impl Foo {
+    extern "rust-call" fn bar() {}
+    //~^ ERROR A function with the "rust-call" ABI must take a single non-self argument
+}
+
+impl Tr for Foo {
+    fn a() {}
+    //~^ ERROR A function with the "rust-call" ABI must take a single non-self argument
+}
+
 fn main () {
     b(10);
+
+    Foo::bar();
+
+    <Foo as Tr>::a();
+    <Foo as Tr>::b();
 }

--- a/src/test/ui/abi/issues/issue-22565-rust-call.rs
+++ b/src/test/ui/abi/issues/issue-22565-rust-call.rs
@@ -5,7 +5,6 @@ extern "rust-call" fn b(_i: i32) {}
 
 trait Tr {
     extern "rust-call" fn a();
-    //~^ ERROR A function with the "rust-call" ABI must take a single non-self argument
 
     extern "rust-call" fn b() {}
     //~^ ERROR A function with the "rust-call" ABI must take a single non-self argument
@@ -19,7 +18,7 @@ impl Foo {
 }
 
 impl Tr for Foo {
-    fn a() {}
+    extern "rust-call" fn a() {}
     //~^ ERROR A function with the "rust-call" ABI must take a single non-self argument
 }
 

--- a/src/test/ui/abi/issues/issue-22565-rust-call.stderr
+++ b/src/test/ui/abi/issues/issue-22565-rust-call.stderr
@@ -4,5 +4,23 @@ error: A function with the "rust-call" ABI must take a single non-self argument 
 LL | extern "rust-call" fn b(_i: i32) {}
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-error: aborting due to previous error
+error: A function with the "rust-call" ABI must take a single non-self argument that is a tuple
+  --> $DIR/issue-22565-rust-call.rs:9:5
+   |
+LL |     extern "rust-call" fn b() {}
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: A function with the "rust-call" ABI must take a single non-self argument that is a tuple
+  --> $DIR/issue-22565-rust-call.rs:16:5
+   |
+LL |     extern "rust-call" fn bar() {}
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: A function with the "rust-call" ABI must take a single non-self argument that is a tuple
+  --> $DIR/issue-22565-rust-call.rs:21:5
+   |
+LL |     extern "rust-call" fn a() {}
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to 4 previous errors
 

